### PR TITLE
[advanced audit] fix invalid status code for hijacker

### LINKED ISSUE #47035: Wrong http code in audit of upgraded connections
The right code is 101, but the advanced audit code is not aware of it if the explicit 101 call is missing.

### PR DESCRIPTION
**Pull request by CaoShuFeng**
Fixes #47035

When using hijacker to take over the connection, the http status code
should be 101 not 200.

PS:
Use "kubectl exec" as an example to review this change.

Part of https://github.com/kubernetes/features/issues/22

### CODE REVIEW

**Comment by CaoShuFeng** [comment]
@sttts

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go [hunk 1]
```diff
@@ -208,6 +208,7 @@ func (f *fancyResponseWriterDelegator) Flush() {
 }
 
 func (f *fancyResponseWriterDelegator) Hijack() (net.Conn, *bufio.ReadWriter, error) {
+	f.processCode(101)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go]: please add comment: // fake an before protocol switch happens

And please use the net/http StatusSwitchingProtocol constant.

Comment (CaoShuFeng) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go]: Done.

**Comment by sttts** [comment]
@k8s-bot ok to test

**Comment by sttts** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by sttts** [comment]
come on @k8s-merge-robot, either tell me what's wrong or approve.

**Comment by sttts** [comment]
@k8s-merge-robot thank you

**Comment by dims** [comment]
👍 LGTM

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go [hunk 2]
```diff
@@ -208,6 +208,8 @@ func (f *fancyResponseWriterDelegator) Flush() {
 }
 
 func (f *fancyResponseWriterDelegator) Hijack() (net.Conn, *bufio.ReadWriter, error) {
+	// fake a response status before protocol switch happens
+	f.processCode(http.StatusSwitchingProtocols)
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go]: not all hijacks switch protocols... for example, streaming logs hijacks

Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go]: Which means that in the case of Hijacking we actually have no clue what will happen. The hijacker might write headers or another code, i.e. manually via "binary" write. The only thing we can do is to introduce another audit stage without any http code (similar to the panic stage). /cc @timstclair @soltysh @CaoShuFeng is it worth it to add that?

Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go]: until we figure that out, please revert this change... setting 101 in non-switching cases is an API break

Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go]: unless I'm misreading this and this is just recording the status for audit

Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go]: It's audit only. We don't have a proper way to express the hijack state. So neither the new nor the old audit event was correct. No need to revert IMO.

Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go]: yeah, that's fine, I misread
